ibm-telemetry/telemetry-js
# Js Scope - Coding Logic

Closes #197, Closes #220, Closes #217

- Adds the programming logic that detects tokens and functions and maps them to the instrumented package, ultimately generating token and function metrics.
- Adds unit testing for all things Js Scope.
- Reworks substitution logic to be able to take in arrays and objects.
- `createSourceFileFromText` function for unit testing ease-of-use
- Fixes significant bug in `getPackageData` that was preventing collection under certain circumstances.

#### Changelog

**New**

- Split substitution logic into new`substituteArray` and `substituteObject` functions, along with a higher-logic `Substitution` class.
- New `getAccessPath` function determines the access path to a function or token (e.g. ['object', 'property', 'token'].
- `deduplicateFunctions` and `deduplicatedTokens` inner functions in `JsScope`.
- New project and sample fixtures for JS scope unit testing.
- Test files containing unit tests and e2e scope for added functions and logic.
- "empty-package-json" project fixture to test edge case.
- Adds test file and cases for `findRelevantSourceFiles`

**Changed**

- Updates `@ibm/telemetry-attributes-js` version to `3.1.0`
- Refines exit condition in `getPackageData` to ignore empty packages.
- Excludes test fixtures from sonarcloud reporting.
- Makes `ComplexValue` readonly.
- Consolidates function-import and token-import matchers to be a combined set of "js-*-import-matcher"s
- Fixes overcollection bug on `JsNamedImportMatcher` & `JsRenamedImportMatcher`
- Consolidates `PropertyAccessExpressionHandler` & `ElementAccessExpressionNodeHandler` to single `AccessExpressionNodeHandler`
- Add `startPos` and `endPos` to `JsToken` & `JsFunction` interfaces (used to track token and function duplication)
- Allow `accessPath` property of `JsToken` & `JsFunction` interfaces to take in `ComplexValue`s (used for sensitive information substitution)
- Remove `readonly` property of `JsAccumulator`'s `imports` (gets rewrittein in `removeIrrelevantImports`)
- Remove `readonly` property of `JsFunctionTokenAccumulator`'s `tokens` and `functions` (gets rewrittein in deduplicate functions)
- `FunctionMetric`:
     - Use new substitution logic
     - Add substitution logic to handle renames, `all` imports and complex values.
     - Add `FUNCTION_MODULE_SPECIFIER`
     - Returns `FUNCTION_ARGUMENT_VALUES` as a space-separated string instead of array.
- `TokenMetric`:
     - Use new substitution logic
     - Add substitution logic to handle renames, `all` imports and complex values.
     - Add `TOKEN_MODULE_SPECIFIER`
- Implement `CallExpressionNodeHandler` logic to catch and build `JsFunction`s
- Implements `IdentifierNodeHandler` logic to catch and build `JsToken`s from identifiers
- Use `substituteObject` function in `ElementMetric`

**Removed**

- src/main/core/anonymize/substitute in favor of `substituteObject` and `substituteArray` functions.

#### Testing / reviewing

All unit tests should pass, test coverage should be above threshold and logic should read seamlessly.

Local Test:
- Pull down this branch
- `npm install`
- `npm run build`
- Go to a local clone of a repository that contains an instrumented package (say [carbon-website](https://github.com/carbon-design-system/carbon-website) or [carbon-platform](https://github.com/carbon-design-system/carbon-platform))
- Update local `telemetry.yml` to include js scope: 
```
collect:
...
  js:
    functions: {}
    tokens:
```
- From inside the instrumented package's directory, run `CI=true node /path-to-telemetry/telemetry-js/dist/collect.js --config=telemetry.yml`, verify log file
